#!/usr/bin/env python3
"""
–ü—Ä–∏–º–µ—Ä –∏–Ω—Ç–µ–≥—Ä–∞—Ü–∏–∏ Python —Å–µ—Ä–≤–∏—Å–æ–≤ —Å Rust –º–∏–∫—Ä–æ—Å–µ—Ä–≤–∏—Å–æ–º —á–µ—Ä–µ–∑ gRPC

–≠—Ç–æ—Ç —Å–∫—Ä–∏–ø—Ç –¥–µ–º–æ–Ω—Å—Ç—Ä–∏—Ä—É–µ—Ç –∫–∞–∫ –∑–∞–º–µ–Ω–∏—Ç—å Python + Neo4j —É–∫–ª–∞–¥–∫—É
–Ω–∞ –≤—ã—Å–æ–∫–æ–ø—Ä–æ–∏–∑–≤–æ–¥–∏—Ç–µ–ª—å–Ω—ã–π Rust —Å–µ—Ä–≤–∏—Å.
"""

import asyncio
import grpc
import logging
import time
from typing import List, Dict, Any

# –ò–º–ø–æ—Ä—Ç —Å–≥–µ–Ω–µ—Ä–∏—Ä–æ–≤–∞–Ω–Ω—ã—Ö protobuf –∫–ª–∞—Å—Å–æ–≤
# –í —Ä–µ–∞–ª—å–Ω–æ–º –ø—Ä–æ–µ–∫—Ç–µ —ç—Ç–∏ —Ñ–∞–π–ª—ã –±—É–¥—É—Ç —Å–≥–µ–Ω–µ—Ä–∏—Ä–æ–≤–∞–Ω—ã –∏–∑ proto/graph_layout.proto
# import graph_layout_pb2
# import graph_layout_pb2_grpc

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

class RustLayoutIntegration:
    """–ò–Ω—Ç–µ–≥—Ä–∞—Ü–∏—è —Å Rust –º–∏–∫—Ä–æ—Å–µ—Ä–≤–∏—Å–æ–º —É–∫–ª–∞–¥–∫–∏ –≥—Ä–∞—Ñ–æ–≤"""
    
    def __init__(self, grpc_endpoint: str = "localhost:50051"):
        self.grpc_endpoint = grpc_endpoint
        self.channel = None
        self.client = None
    
    async def connect(self):
        """–ü–æ–¥–∫–ª—é—á–µ–Ω–∏–µ –∫ Rust gRPC —Å–µ—Ä–≤–∏—Å—É"""
        logger.info(f"üîå –ü–æ–¥–∫–ª—é—á–µ–Ω–∏–µ –∫ Rust —Å–µ—Ä–≤–∏—Å—É –Ω–∞ {self.grpc_endpoint}")
        
        self.channel = grpc.aio.insecure_channel(self.grpc_endpoint)
        # self.client = graph_layout_pb2_grpc.GraphLayoutServiceStub(self.channel)
        
        # –ü—Ä–æ–≤–µ—Ä–∫–∞ –∑–¥–æ—Ä–æ–≤—å—è —Å–µ—Ä–≤–∏—Å–∞
        await self.health_check()
    
    async def health_check(self):
        """–ü—Ä–æ–≤–µ—Ä–∫–∞ –∑–¥–æ—Ä–æ–≤—å—è Rust —Å–µ—Ä–≤–∏—Å–∞"""
        try:
            # –í —Ä–µ–∞–ª—å–Ω–æ–π —Ä–µ–∞–ª–∏–∑–∞—Ü–∏–∏:
            # request = graph_layout_pb2.HealthRequest(service="graph_layout")
            # response = await self.client.GetHealth(request)
            
            logger.info("‚úÖ Rust —Å–µ—Ä–≤–∏—Å —Ä–∞–±–æ—Ç–∞–µ—Ç –Ω–æ—Ä–º–∞–ª—å–Ω–æ")
            return True
        except Exception as e:
            logger.error(f"‚ùå –û—à–∏–±–∫–∞ –ø–æ–¥–∫–ª—é—á–µ–Ω–∏—è –∫ Rust —Å–µ—Ä–≤–∏—Å—É: {e}")
            return False
    
    async def compute_layout_rust(
        self, 
        task_id: str,
        edges: List[Dict[str, Any]] = None
    ) -> Dict[str, Any]:
        """
        –í—ã—á–∏—Å–ª–µ–Ω–∏–µ —É–∫–ª–∞–¥–∫–∏ —á–µ—Ä–µ–∑ Rust —Å–µ—Ä–≤–∏—Å
        
        Args:
            task_id: –ò–¥–µ–Ω—Ç–∏—Ñ–∏–∫–∞—Ç–æ—Ä –∑–∞–¥–∞—á–∏
            edges: –°–ø–∏—Å–æ–∫ —Å–≤—è–∑–µ–π (–µ—Å–ª–∏ None, –∑–∞–≥—Ä—É–∂–∞–µ—Ç—Å—è –∏–∑ Neo4j)
        
        Returns:
            –†–µ–∑—É–ª—å—Ç–∞—Ç —É–∫–ª–∞–¥–∫–∏ —Å –ø–æ–∑–∏—Ü–∏—è–º–∏ –≤–µ—Ä—à–∏–Ω
        """
        logger.info(f"ü¶Ä –ó–∞–ø—É—Å–∫ Rust —É–∫–ª–∞–¥–∫–∏ (–∑–∞–¥–∞—á–∞: {task_id})")
        start_time = time.time()
        
        try:
            # –í —Ä–µ–∞–ª—å–Ω–æ–π —Ä–µ–∞–ª–∏–∑–∞—Ü–∏–∏:
            # 
            # # –ü–æ–¥–≥–æ—Ç–æ–≤–∫–∞ –∑–∞–ø—Ä–æ—Å–∞
            # graph_edges = []
            # if edges:
            #     for edge in edges:
            #         graph_edges.append(graph_layout_pb2.GraphEdge(
            #             source_id=edge['source_id'],
            #             target_id=edge['target_id'],
            #             weight=edge.get('weight', 1.0),
            #             edge_type=edge.get('edge_type', 'BIBLIOGRAPHIC_LINK')
            #         ))
            # 
            # options = graph_layout_pb2.LayoutOptions(
            #     block_width=200.0,
            #     block_height=80.0,
            #     horizontal_gap=40.0,
            #     vertical_gap=50.0,
            #     exclude_isolated_vertices=True,
            #     enable_simd=True,
            #     max_workers=4,
            #     chunk_size=10000
            # )
            # 
            # request = graph_layout_pb2.LayoutRequest(
            #     task_id=task_id,
            #     edges=graph_edges,
            #     options=options
            # )
            # 
            # # –í—ã–ø–æ–ª–Ω–µ–Ω–∏–µ —É–∫–ª–∞–¥–∫–∏
            # response = await self.client.ComputeLayout(request)
            # 
            # if not response.success:
            #     raise Exception(f"Rust —É–∫–ª–∞–¥–∫–∞ –Ω–µ —É–¥–∞–ª–∞—Å—å: {response.error_message}")
            # 
            # # –ö–æ–Ω–≤–µ—Ä—Ç–∞—Ü–∏—è —Ä–µ–∑—É–ª—å—Ç–∞—Ç–∞
            # positions = []
            # for pos in response.positions:
            #     positions.append({
            #         'article_id': pos.article_id,
            #         'layer': pos.layer,
            #         'level': pos.level,
            #         'x': pos.x,
            #         'y': pos.y,
            #         'status': pos.status
            #     })
            
            # –ó–∞–≥–ª—É—à–∫–∞ –¥–ª—è –¥–µ–º–æ–Ω—Å—Ç—Ä–∞—Ü–∏–∏
            positions = [
                {
                    'article_id': f'article_{i}',
                    'layer': i % 5,
                    'level': i // 5,
                    'x': (i % 5) * 240.0,
                    'y': (i // 5) * 130.0,
                    'status': 'placed'
                }
                for i in range(100)
            ]
            
            processing_time = time.time() - start_time
            
            result = {
                'success': True,
                'positions': positions,
                'statistics': {
                    'processing_time_seconds': processing_time,
                    'vertices_processed': len(positions),
                    'algorithm_version': 'rust-0.1.0',
                },
                'metadata': {
                    'server_id': 'rust-server-1',
                    'optimizations_used': ['SIMD', 'Parallel Processing'],
                }
            }
            
            logger.info(
                f"‚úÖ Rust —É–∫–ª–∞–¥–∫–∞ –∑–∞–≤–µ—Ä—à–µ–Ω–∞ –∑–∞ {processing_time:.2f}—Å "
                f"({len(positions)} –ø–æ–∑–∏—Ü–∏–π)"
            )
            
            return result
            
        except Exception as e:
            logger.error(f"‚ùå –û—à–∏–±–∫–∞ Rust —É–∫–ª–∞–¥–∫–∏: {e}")
            return {
                'success': False,
                'error': str(e),
                'positions': [],
                'statistics': {
                    'processing_time_seconds': time.time() - start_time,
                }
            }
    
    async def compute_layout_streaming(self, task_id: str):
        """–ü–æ—Ç–æ–∫–æ–≤–∞—è —É–∫–ª–∞–¥–∫–∞ –¥–ª—è –±–æ–ª—å—à–∏—Ö –≥—Ä–∞—Ñ–æ–≤"""
        logger.info(f"üåä –ó–∞–ø—É—Å–∫ –ø–æ—Ç–æ–∫–æ–≤–æ–π Rust —É–∫–ª–∞–¥–∫–∏ (–∑–∞–¥–∞—á–∞: {task_id})")
        
        # –í —Ä–µ–∞–ª—å–Ω–æ–π —Ä–µ–∞–ª–∏–∑–∞—Ü–∏–∏:
        # request = graph_layout_pb2.LayoutRequest(task_id=task_id, ...)
        # async for chunk in self.client.ComputeLayoutStreaming(request):
        #     yield {
        #         'chunk_id': chunk.chunk_id,
        #         'progress': chunk.progress,
        #         'positions': [pos for pos in chunk.positions],
        #         'is_final': chunk.is_final
        #     }
        
        # –ó–∞–≥–ª—É—à–∫–∞
        for i in range(5):
            await asyncio.sleep(0.1)
            yield {
                'chunk_id': i,
                'progress': (i + 1) / 5.0,
                'positions': [f'chunk_{i}_position_{j}' for j in range(10)],
                'is_final': i == 4
            }
    
    async def get_metrics(self) -> Dict[str, Any]:
        """–ü–æ–ª—É—á–µ–Ω–∏–µ –º–µ—Ç—Ä–∏–∫ –ø—Ä–æ–∏–∑–≤–æ–¥–∏—Ç–µ–ª—å–Ω–æ—Å—Ç–∏"""
        try:
            # request = graph_layout_pb2.MetricsRequest()
            # response = await self.client.GetMetrics(request)
            
            # –ó–∞–≥–ª—É—à–∫–∞
            return {
                'requests_total': 150,
                'requests_success': 145,
                'requests_failed': 5,
                'avg_processing_time': 2.5,
                'memory_usage_mb': 512,
                'cpu_usage_percent': 45.0,
                'vertices_per_second': 2500.0
            }
        except Exception as e:
            logger.error(f"–û—à–∏–±–∫–∞ –ø–æ–ª—É—á–µ–Ω–∏—è –º–µ—Ç—Ä–∏–∫: {e}")
            return {}
    
    async def close(self):
        """–ó–∞–∫—Ä—ã—Ç–∏–µ —Å–æ–µ–¥–∏–Ω–µ–Ω–∏—è"""
        if self.channel:
            await self.channel.close()


async def compare_python_vs_rust():
    """–°—Ä–∞–≤–Ω–µ–Ω–∏–µ –ø—Ä–æ–∏–∑–≤–æ–¥–∏—Ç–µ–ª—å–Ω–æ—Å—Ç–∏ Python vs Rust —É–∫–ª–∞–¥–∫–∏"""
    logger.info("üìä –°—Ä–∞–≤–Ω–µ–Ω–∏–µ –ø—Ä–æ–∏–∑–≤–æ–¥–∏—Ç–µ–ª—å–Ω–æ—Å—Ç–∏ Python vs Rust")
    
    rust_integration = RustLayoutIntegration()
    await rust_integration.connect()
    
    # –¢–µ—Å—Ç –ø—Ä–æ–∏–∑–≤–æ–¥–∏—Ç–µ–ª—å–Ω–æ—Å—Ç–∏
    test_cases = [
        {"name": "–ú–∞–ª–µ–Ω—å–∫–∏–π –≥—Ä–∞—Ñ", "vertex_count": 100},
        {"name": "–°—Ä–µ–¥–Ω–∏–π –≥—Ä–∞—Ñ", "vertex_count": 1000},
        {"name": "–ë–æ–ª—å—à–æ–π –≥—Ä–∞—Ñ", "vertex_count": 10000},
    ]
    
    for test_case in test_cases:
        logger.info(f"üß™ –¢–µ—Å—Ç–∏—Ä–æ–≤–∞–Ω–∏–µ: {test_case['name']}")
        
        # –°–∏–º—É–ª—è—Ü–∏—è Python —É–∫–ª–∞–¥–∫–∏
        python_start = time.time()
        await asyncio.sleep(0.5)  # –ò–º–∏—Ç–∞—Ü–∏—è —Ä–∞–±–æ—Ç—ã Python
        python_time = time.time() - python_start
        
        # Rust —É–∫–ª–∞–¥–∫–∞
        rust_result = await rust_integration.compute_layout_rust(
            task_id=f"test_{test_case['name'].lower().replace(' ', '_')}"
        )
        rust_time = rust_result['statistics']['processing_time_seconds']
        
        speedup = python_time / rust_time if rust_time > 0 else float('inf')
        
        logger.info(f"  Python: {python_time:.3f}—Å")
        logger.info(f"  Rust:   {rust_time:.3f}—Å")
        logger.info(f"  –£—Å–∫–æ—Ä–µ–Ω–∏–µ: {speedup:.1f}x")
        logger.info(f"  –£—Å–ø–µ—Ö: {'‚úÖ' if rust_result['success'] else '‚ùå'}")
    
    await rust_integration.close()


async def integration_example():
    """–ü—Ä–∏–º–µ—Ä –ø–æ–ª–Ω–æ–π –∏–Ω—Ç–µ–≥—Ä–∞—Ü–∏–∏ —Å –∑–∞–º–µ–Ω–æ–π Python –≤–æ—Ä–∫–µ—Ä–∞"""
    logger.info("üîÑ –ü—Ä–∏–º–µ—Ä –∏–Ω—Ç–µ–≥—Ä–∞—Ü–∏–∏ Rust –º–∏–∫—Ä–æ—Å–µ—Ä–≤–∏—Å–∞")
    
    rust_integration = RustLayoutIntegration()
    await rust_integration.connect()
    
    # 1. –ü—Ä–æ–≤–µ—Ä–∫–∞ –∑–¥–æ—Ä–æ–≤—å—è
    health_ok = await rust_integration.health_check()
    if not health_ok:
        logger.error("‚ùå Rust —Å–µ—Ä–≤–∏—Å –Ω–µ–¥–æ—Å—Ç—É–ø–µ–Ω")
        return
    
    # 2. –ü–æ–ª—É—á–µ–Ω–∏–µ –º–µ—Ç—Ä–∏–∫
    metrics = await rust_integration.get_metrics()
    logger.info(f"üìà –ú–µ—Ç—Ä–∏–∫–∏ —Å–µ—Ä–≤–∏—Å–∞: {metrics}")
    
    # 3. –í—ã—á–∏—Å–ª–µ–Ω–∏–µ —É–∫–ª–∞–¥–∫–∏
    result = await rust_integration.compute_layout_rust(
        task_id="integration_test",
        edges=None  # –ó–∞–≥—Ä—É–∑–∏—Ç—Å—è –∏–∑ Neo4j
    )
    
    if result['success']:
        logger.info(f"‚úÖ –£–∫–ª–∞–¥–∫–∞ —É—Å–ø–µ—à–Ω–∞: {len(result['positions'])} –ø–æ–∑–∏—Ü–∏–π")
        logger.info(f"üìä –°—Ç–∞—Ç–∏—Å—Ç–∏–∫–∞: {result['statistics']}")
    else:
        logger.error(f"‚ùå –û—à–∏–±–∫–∞ —É–∫–ª–∞–¥–∫–∏: {result.get('error', 'Unknown')}")
    
    # 4. –ü–æ—Ç–æ–∫–æ–≤–∞—è –æ–±—Ä–∞–±–æ—Ç–∫–∞ (–¥–ª—è –±–æ–ª—å—à–∏—Ö –≥—Ä–∞—Ñ–æ–≤)
    logger.info("üåä –¢–µ—Å—Ç–∏—Ä–æ–≤–∞–Ω–∏–µ –ø–æ—Ç–æ–∫–æ–≤–æ–π –æ–±—Ä–∞–±–æ—Ç–∫–∏")
    async for chunk in rust_integration.compute_layout_streaming("streaming_test"):
        logger.info(f"  –ß–∞–Ω–∫ {chunk['chunk_id']}: {chunk['progress']*100:.1f}%")
        if chunk['is_final']:
            logger.info("  ‚úÖ –ü–æ—Ç–æ–∫–æ–≤–∞—è –æ–±—Ä–∞–±–æ—Ç–∫–∞ –∑–∞–≤–µ—Ä—à–µ–Ω–∞")
    
    await rust_integration.close()


async def main():
    """–ì–ª–∞–≤–Ω–∞—è —Ñ—É–Ω–∫—Ü–∏—è –¥–ª—è –¥–µ–º–æ–Ω—Å—Ç—Ä–∞—Ü–∏–∏ –∏–Ω—Ç–µ–≥—Ä–∞—Ü–∏–∏"""
    logger.info("üöÄ –î–µ–º–æ–Ω—Å—Ç—Ä–∞—Ü–∏—è –∏–Ω—Ç–µ–≥—Ä–∞—Ü–∏–∏ Python ‚Üî Rust")
    
    try:
        await integration_example()
        logger.info("\n" + "="*50)
        await compare_python_vs_rust()
        
        logger.info("\nüéâ –î–µ–º–æ–Ω—Å—Ç—Ä–∞—Ü–∏—è –∑–∞–≤–µ—Ä—à–µ–Ω–∞ —É—Å–ø–µ—à–Ω–æ!")
        
    except Exception as e:
        logger.error(f"‚ùå –û—à–∏–±–∫–∞ –¥–µ–º–æ–Ω—Å—Ç—Ä–∞—Ü–∏–∏: {e}")


if __name__ == "__main__":
    asyncio.run(main())
